### PR TITLE
WIP - support for multiple instances

### PR DESCRIPTION
**Pull request by levb**
#### Summary
Adds server-side of multi-instance support.

- All instance-specific functionality should be working, except the missing instance selectors as listed below. `/jira instance default <URL>` can be used for now to select the instance to use in these contexts.
- `/jira [un]install type <URL>` should work as before, but adds/removes the instances to the list.
- `/jira [dis]connect --instance <URL>` can be used to connect to a specific instance.

###### Changes:
- Fully restored the list of known instances, with an explicit default. `p.GetDefaultInstance(id)` defaults to the explicit default, or the only instance if there's exactly 1 in the list.
- The concept of `CurrentInstance` is replaced by `p.GetDefaultInstance(id)` with an optional id to access a specific instance.
- No more "current instance" caching, all new webhook paths (with `/instance/`) require a trip to the database to fetch the instance list.
- http JSON APIs support passing in `InstanceID` in JSON. Webhooks, callbacks, and user pages support `/instance/XXX` path prefix that identifies the instance (base64-encoded). 
- most commands accept an optional `--instance` flag
- brought in `utils/kvstore` to simplify dealing with instances

###### TODO:
- [ ] webhook URLs - update the command, config screen
- [ ] update `/jira help`
- [x] Data migration from V2: make sure CurrentInstance is added to Known, and if >1 made the default.
- [x] Data migration for V2: create `User` records - lazily?
- [x] Make sure `[dis]connectUser` updates the `User` record.
- [x] Finish updating `GetUserInfo`: [x] WebApp, [x] websockets, [x] `/jira info`
- [x] Bug: deleting an instance should remove user connections - how??? lazily?

###### Tickets to file post-merge:
- [ ] (new ticket) Webapp - add `/jira instance connect` intercept to pop up a list of available instances
- [ ] (new ticket) Webapp - add `/jira instance disconnect` intercept to pop up a list of connected instances (if >1)
- [ ] (new ticket) Webapp - add connected instance selector to Create/Attach
- [ ] Support instance aliases - maybe alias->ID since the instance URLs are stored separately anyway
- [ ] (new ticket) Add user-level `/jira settings default-instance`
- [ ] (new ticket) note to self/TODO: there's some fuzziness with the URL/ID duplication - why? This had something to do with the server URL being set separately, need to review this
- [ ] (new ticket) note to self/TODO: (review PluginVersion - should be set transparently when saving)

#### Ticket Link
Server-side work for #369

### CODE REVIEW

**Comment by levb** [comment]
@mickmister @jfrerich here's the bulk of the server-side work; I'd like to get some early feedback, even if superficial before I get to the webapp part.

**Comment by jfrerich** [comment]
@levb, I am off of set this week and will get to this PR soon!

File: server/instance_server.go [hunk 1]
```diff
@@ -125,52 +113,36 @@ func (jsi jiraServerInstance) GetClient(jiraUser JIRAUser) (client Client, retur
...
-			AccessTokenURL:  jsi.GetURL() + "/plugins/servlet/oauth/access-token",
+			RequestTokenURL: si.GetURL() + "/plugins/servlet/oauth/request-token",
+			AuthorizeURL:    si.GetURL() + "/plugins/servlet/oauth/authorize",
+			AccessTokenURL:  si.GetURL() + "/plugins/servlet/oauth/access-token",
 		},
 		Signer: &oauth1.RSASigner{PrivateKey: rsaKey},
-	}
-
-	return jsi.oauth1Config, nil
+	}, nil
```
[the remark this answers is not in the document]
Comment (levb) [reply server/instance_server.go:147]: I think this runs for all clients, I'll double-check; the RSAKey should be cached still. I need to check.

File: server/instances.go [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	// Notify users we have uninstalled an instance
+	p.API.PublishWebSocketEvent(
+		wSEventInstanceStatus,
+		map[string]interface{}{
+			"instance_installed": false,
+			"instance_type":      "",
```
[the remark this answers is not in the document]
Comment (levb) [reply server/instances.go]: this is changing as we speak

File: server/instance.go [hunk 3]
```diff
@@ -22,23 +29,30 @@ type InstanceCommon struct {
...
-	Alias     string
-	Type      string
-	IsDefault bool
+	InstanceID types.ID
+	Type       InstanceType
+	IsDefault  bool
 }
 
-func newInstanceCommon(p *Plugin, typ string, url types.ID) *InstanceCommon {
+func newInstanceCommon(p *Plugin, instanceType InstanceType, instanceID types.ID) *InstanceCommon {
```
[the remark this answers is not in the document]
Comment (levb) [reply server/instance.go:37]: It literally makes no difference; this sort of implies that it just passes the pointer through. Also, for godoc it's nice to have `New` methods as functions, they end up groupped well.

File: server/command.go [hunk 4]
```diff
@@ -514,64 +514,77 @@ func executeInfo(p *Plugin, c *plugin.Context, header *model.CommandArgs, args .
...
+	case info.IsConnected:
+		resp += fmt.Sprintf("###### Connected to %v Jira instances:\n", info.User.ConnectedInstances.Len())
+	case info.Instances.Len() > 0:
+		resp += fmt.Sprintf("Jira is installed, but you are not connected. Please [connect](%s%s).\n",
+			p.GetPluginURL(), routeUserConnect)
+	default:
+		return p.responsef(header, resp+"\nNo Jira instances installed, please contact your system administrator.")
+	}
+
+	if info.IsConnected {
```
[the remark this answers is not in the document]
Comment (levb) [reply server/command.go:550]: These are the connected instances for the user; the overall list of instances is later.

File: server/instances.go [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+	}
+
+	for _, id := range instances.ValueSet.IDs() {
+		instance := instances.Get(id)
+		if instance.IsDefault {
+			instances.defaultID = id
+			return instance
+		}
+	}
+	instances.defaultID = "not:::a:::url"
```
[the remark this answers is not in the document]
Comment (levb) [reply server/instances.go:68]: Just a guaranteed "0xbaadbeef" - a non-existent key, cached; as opposed to a `""` which is an uninitialized cache value. Just didn't want to do a scan of instances every time, but I'm going to remove the cached pointer and scan; cheap enough, simpler.

File: server/instance_server.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (levb) [reply server/instance_server.go]: Good catch. I moved the public key creation into the OnActivate, and cached in conf

**Comment by levb** [comment]
@aaronrothschild this is basically ready for review - it's missing the (critical?) UI component to select the instance in connect/disconnect/create/attach/subscribe but otherwise should be working.

@mickmister I'd appreciate a through, in-depth review of subs if you remember the code there. I haven't tested it much yet; it _should_ be filtering by instanceID as appropriate. I'd also like to deprecate the "legacy" URLs - have the webhooks always go to /instance/XXX. Need to think through the backwards compatibility - probably ok since migration sets the old instance as "default".

**Comment by levb** [comment]
> connect to frontend logic?

Everything should be in place. I hooked up the existing code (GetInfo, websocket updates, and command hooks). The instance selectors can get all of their data from `.instances` and `.user.connected_instances`. If @mickmister or @jfrerich can quickly mock them up would be great, I was planning on filing them as separate issues once this is merged.

File: server/kv.go [hunk 6]
```diff
@@ -655,3 +440,203 @@ func (store store) OneTimeLoadOauth1aTemporaryCredentials(mmUserId string) (*OAu
...
+	return nil
+}
+
+// MigrateV2User migrates a user record from the V2 data model if needed. It
+// returns an up-to-date User object either way.
+func (p *Plugin) MigrateV2User(mattermostUserID types.ID) (*User, error) {
+	user, err := p.userStore.LoadUser(mattermostUserID)
+	if err != kvstore.ErrNotFound {
+		// return the existing key (or error)
+		return user, err
```
Comment (mickmister) [inline server/kv.go:613]: I'm experiencing an issue here. The user key is not in the kvstore, yet no error is bubbled up from https://github.com/levb/mattermost-plugin-jira/blob/545ff584f0797b5015b4669618997590c30e27c2/server/kv.go#L113 which is necessary for continuing the migration

Comment (mickmister) [reply server/kv.go:613]: Changing this to `if err != kvstore.ErrNotFound && user.MattermostUserID != "" {` seems to work in the meantime for my testing.

**Comment by levb** [comment]
Closing the PR for time being.

**Pull request closed without merging**